modules/host/upload.go

```
@@ -13,53 +16,447 @@ import (
+	h.mu.RUnlock()
+	if !exists {
+		return errors.New("no record of that contract")
+	}
+
+	// need to protect against two simultaneous revisions to the same
+	// contract; this can cause inconsistency and data loss, making storage
+	// proofs impossible
+	//
+	// TODO: DOS vector - the host has locked the obligation even though the
+	// renter has not proven themselves to be the owner of the file contract.
+	obligation.mu.Lock()
+	defer obligation.mu.Unlock()
+
+	// open the file in append mode
+	file, err := os.OpenFile(obligation.Path, os.O_CREATE|os.O_RDWR|os.O_APPEND, 0660)
+	if err != nil {
+		return err
+	}
+
+	// rebuild current Merkle tree
+	tree := crypto.NewTree()
+	err = tree.ReadSegments(file)
+	if err != nil {
+		// Error does not need to be checked when closing the file, already
+		// there have been issues related to the filesystem.
+		_ = file.Close()
+		return err
+	}
+
+	// accept new revisions in a loop. The final good transaction will be
+	// submitted to the blockchain.
+	revisionErr := func() error {
+		for {
+			// allow 5 minutes between revisions
+			err := conn.SetDeadline(time.Now().Add(5 * time.Minute))
+			if err != nil {
+				return err
+			}
+
+			// read proposed revision
+			var revTxn types.Transaction
+			if err = encoding.ReadObject(conn, &revTxn, types.BlockSizeLimit); err != nil {
+				return errors.New("couldn't read revision: " + err.Error())
+			}
+			// an empty transaction indicates completion
+			if revTxn.ID() == (types.Transaction{}).ID() {
+				return nil
+			}
+
+			// allow 5 minutes for each revision
+			err = conn.SetDeadline(time.Now().Add(5 * time.Minute))
+			if err != nil {
+				return err
+			}
+
+			// check revision against original file contract
+			h.mu.RLock()
+			err = h.considerRevision(revTxn, obligation)
+			h.mu.RUnlock()
+			if err != nil {
+				// There is nothing that can be done if there is an error while
+				// writing to a connection.
+				_ = encoding.WriteObject(conn, err.Error())
```

should probably `return` instead of `continue` if the write fails